daemon: use local CiliumNode resource to populate CiliumInternalIP.

tommyp1ckles opened the pull request:
It appears that during recent refactors, restoring cilium addresses from k8s node objects would only use k8s Node types.

However, in order to restore cilium_host router interface IP from k8s, the agent needs to find an IP of type CiliumInternalIP.

This type is only enumerated on CN types, not K8s Nodes so in it's current state all attempts to restore from k8s would return a nil IP.


```release-note
Fix issue where agent attempting to restore local node information (such as cilium_host ip) would fail on k8s fallback method.  
```

tommyp1ckles committed:
daemon: use local CiliumNode resource to populate CiliumInternalIP.

It appears that during recent refactors, restoring cilium addresses from
k8s node objects would only use k8s Node types.

However, in order to restore cilium_host router interface IP from k8s
(the prioritized restore method), the agent needs to find an IP of type CiliumInternalIP.

This type is only enumerated on CN types, not K8s Nodes so in it's
current state all attempts to restore from k8s would return a nil IP.

As well, we've noticed that non-k8s restorations can occasionally
produce unexpected new IPs causing issues when running in vxlan/ipsec
mode due to delay between xfrm state and the router ip being emitted via
apiserver.

Note: Most cilium host restores should succeed on the configuration
based retore which takes precedence over k8s based restore.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

tommyp1ckles committed:
daemon: add unit test for local node init from k8s.

This tests code path where node ip4/ip6 are not configured manually and
thus restoration is attempted from local Node/Cilium node objects.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

meyskens reviewed (approved):
LGTM for me however I do not have much experience with this part of the codebase, might be best to pull in an aditional reviewer

giorio94 commented on daemon/cmd/local_node_sync.go:
```diff
@@ -120,16 +124,45 @@ func (ini *localNodeSynchronizer) getK8sLocalNode(ctx context.Context) (*slim_co
+		ev.Done(nil)
+		if ev.Kind == resource.Upsert {
+			return ev.Object, nil
+		}
```
The wait period in case the CiliumNode does not exist could be reduced by returning immediately once the Sync event is received. At that point, we are sure that the CiliumNode will never be observed, as it could only be created by the agent itself at a later stage. That should also remove the need for the additional timeout configured for the context. Another possibility could be replacing this whole logic with a one-shot get, but that would be slightly worse in terms of performance (assuming that the CiliumLocalNode is also used elsewhere, as the informer would then be shared).

tommyp1ckles replied:
Yeah good call, I reworked this to return immediately in case of sync, otherwise wait for an upsert.

giorio94 commented on daemon/cmd/local_node_sync.go:
```diff
@@ -142,15 +175,21 @@ func (ini *localNodeSynchronizer) initFromK8s(ctx context.Context, node *node.Lo
 	node.Name = parsedNode.Name
+	node.Labels = parsedNode.Labels
+	node.Annotations = parsedNode.Annotations
+	node.Cluster = parsedNode.Cluster
```
I guess these changes, as well as the one at line 153 below, are not intentional, and should be reverted. They are the reason why a couple of unit tests are failing.

giorio94 commented on daemon/cmd/local_node_sync.go:
```diff
@@ -120,16 +124,45 @@ func (ini *localNodeSynchronizer) getK8sLocalNode(ctx context.Context) (*slim_co
+
+	if parsedNode.Name != k8sCiliumNode.Name {
+		return fmt.Errorf("ciliumnode name %q does not match node name %q", k8sCiliumNode.Name, parsedNode.Name)
+	}
```
This check is likely superfluous, given that the LocalNodeResource already enforces that the name matches: https://github.com/tommyp1ckles/cilium/blob/pr%2Ftp%2Frestore-cilium-host-ip-from-cn/daemon/k8s/resources.go#L53-L68

tommyp1ckles replied:
yeah makes sense, ill remove this.

giorio94 commented on daemon/cmd/local_node_init_test.go:
You may want to merge this file (and possibly part of the content) with `local_node_sync_test.go`.

giorio94 commented on daemon/cmd/local_node_sync.go:
```diff
@@ -52,7 +52,7 @@ func (ini *localNodeSynchronizer) InitLocalNode(ctx context.Context, n *node.Loc
 		return err
 	}
-
+	// Deprecated: Restoring node information from k8s is deprecated as of v1.16.0 and should be removed in v1.17.0.
```
This doesn't look correct to me. While I agree that CiliumInternalIPs cannot be restored from the Node object, this logic also takes care of initializing the LocalNodeStore with node Internal/External IP addresses as well as labels, annotations and a couple of extra fields, whose source of trust is the kubernetes Node object itself. This pieces of information are then eventually propagated into the corresponding CiliumNode object by the [node discovery logic](https://github.com/tommyp1ckles/cilium/blob/pr%2Ftp%2Frestore-cilium-host-ip-from-cn/pkg/nodediscovery/nodediscovery.go#L231-L236).

tommyp1ckles replied:
You're right,  I think what I intended to say here is that we want to deprecate restoring cilium_host IP from k8s.   Saying we want to deprecate restoring from k8s is unfeasible and too broad.

tommyp1ckles replied:
Going to remove this commit from now, I probably actually want to discuss this more.

giorio94 reviewed (changes requested):
Thanks for fixing this! I've left a few comments inline.

tommyp1ckles committed:
test/controlplane: add field filterlist case for ciliumnodelist.

This fixes panic in controlplane tests introduced by previous commits
related to CiliumNode Resource[T].

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

giorio94 commented on daemon/cmd/local_node_sync.go:
```diff
@@ -120,6 +122,28 @@ func (ini *localNodeSynchronizer) getK8sLocalNode(ctx context.Context) (*slim_co
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		return nil
```
Nit. This explicit check is not necessary because the events channel is closed when the context is canceled.

giorio94 reviewed (approved):
Looks good to me, thanks!

I've just left a non-blocking nit inline (feel free to ignore it given that CI seems practically green).

christarazi reviewed (approved):
ci-structure LGTM
